core/src/main/java/org/elasticsearch/common/unit/ByteSizeValue.java

```diff
@@ -20,28 +20,40 @@
 package org.elasticsearch.common.unit;
 
 import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.Version;
 import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.logging.DeprecationLogger;
+import org.elasticsearch.common.logging.Loggers;
 
 import java.io.IOException;
 import java.util.Locale;
 import java.util.Objects;
 
 public class ByteSizeValue implements Writeable, Comparable<ByteSizeValue> {
+    private static final DeprecationLogger DEPRECATION_LOGGER = new DeprecationLogger(Loggers.getLogger(ByteSizeValue.class));
 
     private final long size;
     private final ByteSizeUnit unit;
 
     public ByteSizeValue(StreamInput in) throws IOException {
-        size = in.readVLong();
+        if (in.getVersion().before(Version.V_6_2_0)) {
+            size = in.readVLong();
+        } else {
+            size = in.readZLong();
+        }
         unit = ByteSizeUnit.BYTES;
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
-        out.writeVLong(getBytes());
+        if (out.getVersion().before(Version.V_6_2_0)) {
+            out.writeVLong(getBytes());
+        } else {
+            out.writeZLong(getBytes());
```

Comment:
I think we can do better here? Write out the bytes and write out the unit and read them on de-serialization to preserve the unit?

Reply:
Agreed